src/main/java/bio/terra/service/configuration/StatusService.java

```diff
@@ -0,0 +1,95 @@
+import bio.terra.model.RepositoryStatusModelSystems;
+import bio.terra.service.dataset.DatasetDao;
+import bio.terra.service.iam.IamProviderInterface;
+import org.broadinstitute.dsde.workbench.client.sam.model.SystemStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DataAccessException;
+import org.springframework.stereotype.Service;
+
+@Service
+public class StatusService {
+
+    private final Logger logger = LoggerFactory.getLogger(StatusService.class);
+    private final ConfigurationService configurationService;
+    private final DatasetDao datasetDao;
+    private final IamProviderInterface iamProviderInterface;
+
+    @Autowired
+    public StatusService(
+        ConfigurationService configurationService,
+        DatasetDao datasetDao,
+        IamProviderInterface iamProviderInterface) {
+        this.configurationService = configurationService;
+        this.datasetDao = datasetDao;
+        this.iamProviderInterface = iamProviderInterface;
+    }
+
+    public RepositoryStatusModel getStatus() {
+        RepositoryStatusModel statusModel = new RepositoryStatusModel();
+
+        if (configurationService.testInsertFault(ConfigEnum.LIVENESS_FAULT)) {
+            logger.info("LIVENESS_FAULT insertion - failing status response");
+            statusModel.setOk(false);
+            return statusModel;
+        }
+
+        statusModel.putSystemsItem("Postgres", postgresStatus());
+        statusModel.putSystemsItem("Sam", samStatus());
+
+        // if critical system is down, then isOk = false
+        statusModel.setOk(statusModel.getSystems().values().stream()
+            .noneMatch(sys -> sys.isCritical() && !sys.isOk()));
+
+        return statusModel;
+    }
+
+    private RepositoryStatusModelSystems postgresStatus() {
+        // test by inserting fault
+        if (configurationService.testInsertFault(ConfigEnum.CRITICAL_SYSTEM_FAULT)) {
+            logger.info("CRITICAL_SYSTEM_FAULT inserted for test - setting postgres system status to failing");
+            return new RepositoryStatusModelSystems()
+                .ok(false)
+                .critical(true)
+                .message("CRITICAL_SYSTEM_FAULT inserted for test");
+        }
+
+        Boolean dbStatus = false;
+        String Msg;
+        try {
+            datasetDao.probeDatabase();
+            dbStatus = true;
+            Msg = "Successfully queried database.";
+        } catch (DataAccessException ex) {
+            Msg = "Failed to query database: " + ex.getMessage();
+            logger.error("Failed to query database. Ex: {}", ex);
+        }
+        RepositoryStatusModelSystems databaseSystem = new RepositoryStatusModelSystems()
+            .ok(dbStatus)
+            .critical(true)
+            .message(Msg);
+        return databaseSystem;
+    }
+
+    private RepositoryStatusModelSystems samStatus() {
```

Comment:
Since `postgresStatus` and `samStatus` are largely the same, I wonder if it would be useful to use the Java functional interface to make this extensible. Something like:

```
// need to add this to imports
// a Supplier takes in no inputs and produces a single output
import java.util.function.Supplier;

private boolean checkPostgres() {
    // omitted code here
}

private boolean samStatus() {
    // omitted code here
}

// additional services checks we might add in the future omitted

private RepositoryStatusModelSystems checkStatus(String statusFnName, Supplier <Boolean> statusFn) {
    Boolean serviceStatus = false;
    String msg;
    try {
        serviceStatus = statusFn.get();
        msg = statusFnName + " succeeded";
    } catch (Exception ex) {
        logger.error("Error during {}, exception was: {}", statusFnName, ex);
        msg = ex.getMessage();
    }
    RepositoryStatusModelSystems statusModel = new RepositoryStatusModelSystems()
        .ok(serviceStatus)
        .critical(true)
        .message(msg);
    return statusModel;
}

// example use:
// checkStatus("checkPostgres", StatusService::checkPostgres)
```

Reply:
That would be cleaner. The main issue I see is that I need both the status boolean and the message from the Sam call. I tried the following, but it relies on passing in a function that can throw an exception which doesn't appear to be allowed. 
![image](https://user-images.githubusercontent.com/13254229/98102006-046f5c00-1e61-11eb-928a-791e0d9d678d.png)


``` 
    private RepositoryStatusModelSystems samStatus() throws ApiException {
        SystemStatus samStatusModel = iamProviderInterface.samStatus();

        RepositoryStatusModelSystems samSystem = new RepositoryStatusModelSystems()
            .ok(samStatusModel.getOk())
            .message(samStatusModel.getSystems().toString());
        return samSystem;
    }

    private RepositoryStatusModelSystems checkStatus(String statusFnName, Supplier<RepositoryStatusModelSystems> statusFn, boolean isCritical) {
        RepositoryStatusModelSystems statusModel;
        try {
            statusModel = statusFn.get();
            statusModel.setCritical(isCritical);
        } catch (Exception ex) {
            logger.error("Error during {}, exception was: {}", statusFnName, ex);
            statusModel = new RepositoryStatusModelSystems()
                .ok(false)
                .critical(isCritical)
                .message(ex.getMessage());
        }
        return statusModel;
    }
```

Reply:
In this case, instead of `Supplier`, you'll need to use `Callable`. The critical piece would look like this instead:

```
import java.util.concurrent.Callable;

private RepositoryStatusModelSystems checkStatus(
        String statusFnName,
        Callable<RepositoryStatusModelSystems> statusFn,
        boolean isCritical) {
        RepositoryStatusModelSystems statusModel;
        try {
            statusModel = statusFn.call();
            statusModel.setCritical(isCritical);
        } catch (Exception ex) {
            logger.error("Error during {}, exception was: {}", statusFnName, ex);
            statusModel = new RepositoryStatusModelSystems()
                .ok(false)
                .critical(isCritical)
                .message(ex.getMessage());
        }
        return statusModel;
    }
```

Reply:
This is no longer need with the re-org, but I appreciate these thoughtful suggestions!